internal/provider/vdcg/firewall_resource.go

```diff
@@ -0,0 +1,345 @@
+package vdcg
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+
+	supertypes "github.com/FrangipaneTeam/terraform-plugin-framework-supertypes"
+
+	"github.com/orange-cloudavenue/cloudavenue-sdk-go/pkg/urn"
+	v1 "github.com/orange-cloudavenue/cloudavenue-sdk-go/v1"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/client"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/metrics"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/provider/common/mutex"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                = &FirewallResource{}
+	_ resource.ResourceWithConfigure   = &FirewallResource{}
+	_ resource.ResourceWithImportState = &FirewallResource{}
+)
+
+// NewFirewallResource is a helper function to simplify the provider implementation.
+func NewFirewallResource() resource.Resource {
+	return &FirewallResource{}
+}
+
+// FirewallResource is the resource implementation.
+type FirewallResource struct {
+	client   *client.CloudAvenue
+	vdcGroup *v1.VDCGroup
+}
+
+// Init Initializes the resource.
+func (r *FirewallResource) Init(ctx context.Context, rm *FirewallModel) (diags diag.Diagnostics) {
+	var err error
+
+	idOrName := rm.VDCGroupName.Get()
+	if rm.VDCGroupID.IsKnown() {
+		// Use the ID
```

Comment:
remove this part if the schema is set with the pattern validator: && urn.IsVDCGroup(rm.VDCGroupID.Get())